antenna/configlib.py

```diff
@@ -0,0 +1,323 @@
+    # Special case bool so that we can explicitly give bool values otherwise
+    # all values would be True since they're non-empty strings.
+    if parser is bool:
+        return parse_bool
+    return parser
+
+
+class ListOf(object):
+    def __init__(self, parser, delimiter=','):
+        self.sub_parser = parser
+        self.delimiter = delimiter
+
+    def __call__(self, value):
+        parser = get_parser(self.sub_parser)
+        return [parser(token) for token in value.split(self.delimiter)]
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+class ConfigOverrideEnv(object):
+    """Override configuration layer for testing"""
+    def get(self, key, namespace=''):
+        if namespace:
+            key = '%s_%s' % (namespace, key)
+
+        key = key.upper()
+
+        for env in reversed(_CONFIG_OVERRIDE):
+            if key in env:
+                return env[key]
+        return NO_VALUE
+
+
+class ConfigDictEnv(object):
+    """dict-based configuration layer
+
+    Namespace is prefixed, so key "foo" in namespace "bar" is ``FOO_BAR``.
+
+    """
+    def __init__(self, cfg):
+        self.cfg = cfg
+
+    def get(self, key, namespace=''):
+        if namespace:
+            key = '%s_%s' % (namespace, key)
+        key = key.upper()
+
+        if key in self.cfg:
+            return self.cfg[key]
+        return NO_VALUE
+
+
+class ConfigOSEnv(object):
+    """os.environ derived configuration layer
+
+    Namespace is prefixed, so key "foo" in namespace "bar" is ``FOO_BAR`` in
+    the ``os.environ``.
+
+    """
+    def get(self, key, namespace=''):
+        if namespace:
+            key = '%s_%s' % (namespace, key)
+
+        key = key.upper()
+        if key in os.environ:
+            return os.environ[key]
+
+        return NO_VALUE
+
+
+class ConfigIniEnv(object):
+    """.ini style configuration layer
+
+    Namespace is a config section. So "foo" in namespace "bar" is::
+
+        [bar]
+        foo=someval
+
+    """
+    def __init__(self, fn):
+        self._parser = ConfigParser()
+        self._parser.readfp(open(fn, 'r'))
+
+    def get(self, key, namespace='main'):
+        if self._parser.has_option(namespace, key):
+            return self._parser.get(namespace, key)
+        return NO_VALUE
+
+
+class ConfigManager(object):
+    """Manages multiple configuration environment layers"""
+    def __init__(self):
+        self.envs = self._initialize_environments()
+
+    def _initialize_environments(self):
+        # Add environments in order of precedence--first is the most important!
+        envs = []
+
+        # Add override environment
+        envs.append(ConfigOverrideEnv())
+
+        # Add OS environment
+        envs.append(ConfigOSEnv())
+
+        # Add .ini environment
+        fn = os.environ.get('ANTENNA_INI', 'antenna.ini')
+        if not os.sep in fn:
+            fn = os.path.join(os.getcwd(), fn)
+        if os.path.exists(fn):
+            envs.append(ConfigIniEnv(fn))
```

Comment:
A missing `else` here might cause confusion. If I set `ANTENNA_INI=antenn.ini` (note typo) it's going to give up because it can't find `antenn.ini`.